optuna/multi_objective/_hypervolume/wfg.py

```diff
@@ -0,0 +1,86 @@
+    : 86-95.`.
+    """
+
+    def __init__(self) -> None:
+        self._reference_point = None
+
+    def compute(self, solution_set: np.ndarray, reference_point: np.ndarray) -> float:
+        self._reference_point = reference_point
+        return self._compute_rec(solution_set)
+
+    def _compute_rec(self, solution_set: np.ndarray) -> float:
+        assert self._reference_point is not None
+        n_points = solution_set.shape[0]
+
+        if n_points == 1:
+            return _compute_2points_volume(solution_set[0], self._reference_point)
+        elif n_points == 2:
+            volume = 0.0
+            volume += _compute_2points_volume(solution_set[0], self._reference_point)
+            volume += _compute_2points_volume(solution_set[1], self._reference_point)
+            intersection = self._reference_point - np.maximum(solution_set[0], solution_set[1])
+            volume -= np.prod(intersection)
+
+            return volume
+
+        solution_set = solution_set[solution_set[:, 0].argsort()]
+
+        # n_points >= 3
+        volume = 0.0
+        for i in range(n_points):
+            volume += self._compute_exclusive_hv(solution_set[i], solution_set[i + 1 :])
+        return volume
+
+    def _compute_exclusive_hv(self, point: np.ndarray, solution_set: np.ndarray) -> float:
+        volume = _compute_2points_volume(point, self._reference_point)
+        limited_solution_set = self._limit(point, solution_set)
+        n_points_of_s = limited_solution_set.shape[0]
+        if n_points_of_s == 1:
+            volume -= _compute_2points_volume(limited_solution_set[0], self._reference_point)
+        elif n_points_of_s > 1:
+            volume -= self._compute_rec(limited_solution_set)
+        return volume
+
+    @staticmethod
+    def _limit(point: np.ndarray, solution_set: np.ndarray) -> np.ndarray:
+        n_points_of_s = solution_set.shape[0]
+        dim = point.shape[0]
+        limited_solution_set = []
+
+        for i in range(n_points_of_s):
+            if _dominates_or_equal(point, solution_set[i]):
+                return point.reshape((1, dim))
+            limited_solution_set.append(np.maximum(solution_set[i], point))
+        limited_solution_set = np.asarray(limited_solution_set).reshape(
+            (len(limited_solution_set), dim)
+        )
+
+        # Return only pareto optimal points for computational efficiency.
+        if n_points_of_s <= 1:
+            return limited_solution_set
+        else:
+            # Assume limited_solution_set is sorted by its 0th dimension.
+            returned_limited_solution_set = [limited_solution_set[0]]
+            left = 0
+            right = 1
+            while left < right < n_points_of_s:
+                if not _dominates_or_equal(
+                    limited_solution_set[right], limited_solution_set[left]
+                ):
+                    left = right
+                    returned_limited_solution_set.append(limited_solution_set[left])
```

Comment:
These lines are a bit difficult for me to understand, and I want a comment here too.
Intuitively, if `right` is not dominated by `left`, `right` is a part of the Pareto front.

And also, I'm not sure if this logic always extracts the Pareto optimal solutions.
What happens if the `right` dominates `left`?

```suggestion
                if not _dominates_or_equal(
                    limited_solution_set[right], limited_solution_set[left]
                ):
                    # `right` is not dominated by `left`, and it corresponds to a Pareto optimal point.
                    returned_limited_solution_set.append(limited_solution_set[right])
                    left = right
```

Reply:
For me, the negation of `_dominates_or_equal` is a bit confusing. I prefer the following style, but this is not a request for the change.
```python
                if (limited_solution_set[left] > limited_solution_set[right]).any():
                    # If the `right` is not dominated by `left`
```

Reply:
Thanks for the comment.
>And also, I'm not sure if this logic always extracts the Pareto optimal solutions.
What happens if the right dominates left?

Yes, if the `right` dominates `left`, we have to consider a lot of cases to decide Pareto optimal points. However, the input `solution_set` of `_limit()` method is assumed to be sorted in `_compute_rec()` method (Line 38). Therefore, we don't have to consider such cases.

I left a comment about those points.

Reply:
Maybe I'm wong, but I think `right` can dominate `left` since the `solution_set` is only sorted by 0-th dimension. It means the order of 1st and later dimensions are undefined. For example, if we have two 2-dimensional vectors like `(0, 1)` and `(0, 2)`, the `right` can be `(0, 1)`, which dominates `left=(0, 2)`. In that case, `left` seems to be included even if it is not a Pareto optimal point. The following is an example input:

```python
>>> _limit(np.array([0, 0, 0]), np.array([[0, 1, 2], [0, 2, 2]]))
array([[0, 1, 2]])
>>> _limit(np.array([0, 0, 0]), np.array([[0, 2, 2], [0, 1, 2]]))
array([[0, 2, 2],
       [0, 1, 2]])
```


This phenomenon actually happens if we slightly modified a test case.

<details>
<summary>For example, if we add one more `(0, 1)` to `test_wfg_2d`:</summary>

```python
def test_wfg_2d() -> None:
    for n in range(2, 30):
        r = n * np.ones(2)
        s = np.asarray([[n - 1 - i, i] for i in range(n)])
        for i in range(n + 1):
            s = np.vstack((s, np.asarray([i, n - i])))
        s = np.vstack(np.asarray((s, [0, 1])))  # This line is added.
        np.random.shuffle(s)
        v = optuna.multi_objective._hypervolume.WFG().compute(s, r)
        assert v == n * n - n * (n - 1) // 2
```

Output 1:
```console
Sorted solution set [[0 1]
 [0 1]
 [0 2]
 [1 0]
 [1 1]
 [2 0]]
Point in _limit [0 1]
Solution_set in _limit [[0 1]
 [0 2]
 [1 1]
 [1 1]
 [2 1]]
Limited solution set [array([0, 1])]  # expected
```

Output 2:
```console
Sorted solution set [[0 1]
 [0 2]
 [0 1]
 [1 0]
 [1 1]
 [2 0]]
Point in _limit [0 1]
Solution set in _limit  [[0 2]
 [0 1]
 [1 1]
 [1 1]
 [2 1]]
Limited solution set [array([0, 2]), array([0, 1])]
```
</details>

To fix this case, we'll need to consider all dimensions when we sort the `solution_set` (c.f., we can use `np.lexsort(np.rot90(solution_set))` as mentioned [here](https://stackoverflow.com/questions/38277143/sort-2d-numpy-array-lexicographically)). But, I think it's OK to accept such edge cases since we don't require exact Pareto front here.

In addition to that, I noticed that the current WFG algorithm may not consider the duplication of points. I think it can be seen if the objective values are integer. What do you think of it?

Reply:
I agree with you. Because the performance degradation will be small due to the miss of "complete" Pareto optimal set in the `_limit()` method, I will fix only the comment in the source code. Thank you for your insightful reviews.